### PR TITLE
Remove dead code in System.Private.Xml

### PR DESCRIPTION
**Pull request by sepidehMS**
Unreachable implementations inside System.Private.Xml were discovered by BclRewriter and removed with help of the CodeRemover tool. ~~There were also many public types in `System.Xml.Xsl.Runtime` namespace that were not exposed in the Xml contracts. All of those were made internal.~~ (Related issue: https://github.com/dotnet/corefx/issues/16825)

System.Private.Xml contains implementations from both System.Xml.dll and System.Data.SqlXml.dll on Desktop. Here's how the sizes look now:

2,657,952 System.XML.dll
734,416 System.Data.SqlXml.dll
3,534,336 System.Private.Xml.dll 

cc: @danmosemsft @weshaggard

### CODE REVIEW

**Comment by sepidehMS** [comment]
Something went wrong, some of my changes are not included here. Will update with fix.

**Comment by weshaggard** [comment]
When you did the BCLRewriter step how did you determine the public surface area to root? Also did you compare your results with https://github.com/dotnet/corefx/issues/17905?

**Comment by danmosemsft** [comment]
Used modelgen.exe to configure bclrewriter

**Comment by weshaggard** [comment]
>Used modelgen.exe to configure bclrewriter

Yes but what did you run modelgen against? What set of public contracts?

**Comment by danmosemsft** [comment]
I can't remember, it could have been the current implementation. Which would mean internalsvisibleto would not have been respected guess

**Comment by weshaggard** [comment]
Ideally the model file would be generated from the public API contracts and not the current implementation because the implementation could potentially have a lot more public APIs then the contract that we don't need to keep.

As for internalsvisibileto I'm not worried about those I don't believe we have many of those outside of CoreLib and if we do I want to find them and eliminate them :)

**Comment by danmosemsft** [comment]
@sepidehMS it would make sense to run ILLink after your chnages (instructions in https://github.com/dotnet/corefx/issues/17905) and see if it finds anything else.

File: src/System.Private.Xml/src/System/Xml/Core/XmlEncodedRawTextWriterAsync.cs [hunk 1]
```diff
@@ -258,7 +258,7 @@ protected internal override Task WriteEndAttributeAsync()
 
             if (trackTextContent && inTextContent != false) { ChangeTextContentMark(false); }
 
-			// TODO: check what does this do and remove/keep
+            // TODO: check what does this do and remove/keep
 #if SILVERLIGHT
```
Comment (stephentoub) [inline src/System.Private.Xml/src/System/Xml/Core/XmlEncodedRawTextWriterAsync.cs]: Can we remove all of this SILVERLIGHT code?

File: src/System.Private.Xml/src/System/Xml/Core/XmlTextReaderImplAsync.cs [hunk 2]
```diff
@@ -2706,9 +2706,10 @@ private Task ParseEndElementAsync_CheckEndTag(int nameLen, NodeData startTagNode
                 {
 #if DEBUG
 #if !SILVERLIGHT
```
Comment (stephentoub) [inline src/System.Private.Xml/src/System/Xml/Core/XmlTextReaderImplAsync.cs]: Similarly, assuming we're never compiling with SILVERLIGHT, can we clean up these ifdefs?

File: src/System.Private.Xml/src/System/Xml/DiagnosticsSwitches.cs [hunk 3]
```diff
@@ -7,7 +7,8 @@ namespace System.Xml
     using System.Diagnostics;
 
 #if DEBUG
-    public static class DiagnosticsSwitches {
+    public static class DiagnosticsSwitches
```
Comment (stephentoub) [inline src/System.Private.Xml/src/System/Xml/DiagnosticsSwitches.cs]: Does this ever actually need to be public, or could this debug-specific code be removed?

File: src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs [hunk 4]
```diff
@@ -22,7 +22,7 @@ public class XmlDocument : XmlNode
         private DomNameTable _domNameTable; // hash table of XmlName
         private XmlLinkedNode _lastChild;
         private XmlNamedNodeMap _entities;
-		// TODO: _htElementIdMap can be Dictionary<string, List<WeakReference<XmlElement>>>
+        // TODO: _htElementIdMap can be Dictionary<string, List<WeakReference<XmlElement>>>
```
Comment (stephentoub) [inline src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs]: Are there issues tracking these TODOs?

File: src/System.Private.Xml/src/System/Xml/XPath/Internal/CacheChildrenQuery.cs [hunk 5]
```diff
@@ -92,10 +92,12 @@ public override XPathNavigator Advance()
                         continue;
                     }
                 }
-				// TODO: this code was previously removed
+                // TODO: this code was previously removed
 #if DEBUG
```
Comment (danmosemsft) [inline src/System.Private.Xml/src/System/Xml/XPath/Internal/CacheChildrenQuery.cs]: >if DEBUG [](start = 1, length = 8)

I would delete this #if DEBUG. It looks like temporary code for some VS bug investgiation. And the set of _lastNode only in debug is worrying.

File: src/System.Private.Xml/src/System/Xml/XmlResolver.cs [hunk 6]
```diff
@@ -53,7 +53,7 @@ public virtual Uri ResolveUri(Uri baseUri, string relativeUri)
                 // relative base Uri
                 if (!baseUri.IsAbsoluteUri)
                 {
-					// TODO: should this be SILVERLIGHT or desktop behavior?
+                    // TODO: should this be SILVERLIGHT or desktop behavior?
 #if SILVERLIGHT
```
Comment (danmosemsft) [inline src/System.Private.Xml/src/System/Xml/XmlResolver.cs]: More #if SILVERLIGHT to delete. And the todo.

**Comment by danmosemsft** [comment]
>         /// TODO: Need to make this public    

remove comment

---
Refers to: src/System.Private.Xml/src/System/Xml/Schema/XmlSchemaSet.cs:269 in 1b10f26. [](commit_id = 1b10f26d678c1aa8143233af72f215e286b2ac0e, deletion_comment = False)

File: src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs (hunk 4 above)
Comment (danmosemsft) [inline src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs]: This one can be deleted as it has no value.

---
In reply to: [110308973](https://github.com/dotnet/corefx/pull/18042#discussion_r110308973) [](ancestors = 110308973)

Comment (danmosemsft) [inline src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs]: I looked through the others and none seem worth issues. I assume most carried over from years ago.

---
In reply to: [110315858](https://github.com/dotnet/corefx/pull/18042#discussion_r110315858) [](ancestors = 110315858,110308973)

**Comment by danmosemsft** [comment]
> #if !SILVERLIGHT

#if !SILVERIGHT can be removed also of course

---
Refers to: src/System.Private.Xml/src/System/Xml/Schema/XsdDateTime.cs:247 in 1b10f26. [](commit_id = 1b10f26d678c1aa8143233af72f215e286b2ac0e, deletion_comment = False)

File: src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs (hunk 4 above)
Comment (danmosemsft) [inline src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs]: None of the others seem worthy of an issue. Most look like notes from years ago.

**Comment by danmosemsft** [comment]
Looks good save comments above and the ILMerge suggestion. Also making builds and tests pass.. Thanks @sepidehMS !

**Comment by danmosemsft** [comment]
Occurs to me that since most of what you deleted wasn't unnecessary files, we still haven't explained the mysteriuos size growth over desktop. We've alleviated it, though!

File: src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs]: > None of the others seem worthy of an issue. Most look like notes from years ago.

Can we delete them then?  TODOs should represent things to do, and things to do should have issues.  If there's nothing to do, we should be able to delete the TODO, or at the very least change it to a normal comment if it's just descriptive.

File: src/System.Private.Xml/src/System/Xml/BinaryXml/SqlUtils.cs [hunk 7]
```diff
@@ -65,11 +65,20 @@ public BinXmlSqlDecimal(byte[] data, int offset, bool trim)
             byte b = data[offset];
             switch (b)
             {
-                case 7: m_bLen = 1; break;
-                case 11: m_bLen = 2; break;
-                case 15: m_bLen = 3; break;
-                case 19: m_bLen = 4; break;
-                default: throw new XmlException(SR.XmlBinary_InvalidSqlDecimal, (string[])null);
+                case 7:
```
Comment (krwq) [inline src/System.Private.Xml/src/System/Xml/BinaryXml/SqlUtils.cs]: this looked more readable before the change

File: src/System.Private.Xml/src/System/Xml/BinaryXml/SqlUtils.cs [hunk 8]
```diff
@@ -626,7 +635,7 @@ internal abstract class BinXmlDateTime
     {
         private const int MaxFractionDigits = 7;
 
-        internal static int[] KatmaiTimeScaleMultiplicator = new int[8] {
+        static internal int[] KatmaiTimeScaleMultiplicator = new int[8] {
```
Comment (krwq) [inline src/System.Private.Xml/src/System/Xml/BinaryXml/SqlUtils.cs]: shouldn't access modifiers come first?

**Comment by krwq** [comment]
I don't think I like the stylistic changes in this PR

**Comment by sepidehMS** [comment]
@weshaggard I used the current implementation to get the model.xml. But I also diffed the public surface area against System.Xml.dll in netstandard. Turned out System.Private.Xml.dll has some public types in `System.Xml.Xsl.Runtime` namespace that are not exposed in Xml contracts. These types/APIs are loaded at runtime and making them internal breaks functionality of Xslt. So, I'm leaving those as public as seems like they were intentionally made public but not browsable.

I did a diff against https://github.com/dotnet/corefx/issues/17905 and there are a couple of APIs that we didn't find with BclRewriter. As @danmosemsft suggested, I'm going to run ILLink on top of my changes and remove anything that it finds.

File: src/System.Private.Xml/src/System/Xml/Core/XmlEncodedRawTextWriterAsync.cs (hunk 1 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System/Xml/Core/XmlEncodedRawTextWriterAsync.cs]: Will do.

File: src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs (hunk 4 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System/Xml/Dom/XmlDocument.cs]: Will do a pass on TODOs and remove anything that's no longer needed.

File: src/System.Private.Xml/src/System/Xml/XPath/Internal/CacheChildrenQuery.cs (hunk 5 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System/Xml/XPath/Internal/CacheChildrenQuery.cs]: Will do.

File: src/System.Private.Xml/src/System/Xml/BinaryXml/SqlUtils.cs (hunk 7 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System/Xml/BinaryXml/SqlUtils.cs]: I ran CodeFormatter here to make sure running CodeRemover didn't leave the code in bad style. Isn't the format given by CodeFormatter what we prefer? If everyone agrees it's not necessary, I will skip all the style changes in this PR.

File: src/System.Private.Xml/src/System/Xml/BinaryXml/SqlUtils.cs (hunk 8 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System/Xml/BinaryXml/SqlUtils.cs]: Similarly, the preferred style in CodeFormatter.

File: src/System.Private.Xml/src/System/Xml/BinaryXml/SqlUtils.cs (hunk 7 above)
Comment (danmosemsft) [inline src/System.Private.Xml/src/System/Xml/BinaryXml/SqlUtils.cs]: I think the "team consensus" is to follow CodeFormatter (I'm not aware of any files we agreed to never run it on)

File: src/System.Private.Xml/src/System/Xml/BinaryXml/SqlUtils.cs (hunk 8 above)
Comment (sepidehMS) [inline src/System.Private.Xml/src/System/Xml/BinaryXml/SqlUtils.cs]: Actually, I think I was wrong. I checked CodeFormatter with a small csproj and it seems it doesn't have a preference over the order of this kind. This must be a change I accidentally made myself. Will fix.

**Comment by weshaggard** [comment]
>@weshaggard I used the current implementation to get the model.xml. But I also diffed the public surface area against System.Xml.dll in netstandard. Turned out System.Private.Xml.dll has some public types in System.Xml.Xsl.Runtime namespace that are not exposed in Xml contracts. These types/APIs are loaded at runtime and making them internal breaks functionality of Xslt. So, I'm leaving those as public as seems like they were intentionally made public but not browsable.

Sounds reasonable. For the Xsl APIs do we have another contract that exposes those APIs to the developers? If not then why do we need them?

**Comment by sepidehMS** [comment]
> For the Xsl APIs do we have another contract that exposes those APIs to the developers? If not then why do we need them?

These APIs are currently not exposed in any of the Xml contracts. Is there a reason why these types were not included in NETStandard's public surface area? They used to be exposed in System.Data.SqlXml.dll. I believe we need to expose them in System.Xml.ReaderWriter's contract, where the other Xsl types are exposed right now. @weshaggard

**Comment by weshaggard** [comment]
Sstem.Data.SqlXml.dll is not part of netstandard as it is a Sql specific implementation of xml. Can you help me understand why they need to be in System.Private.Xml as opposed to an independent library?

**Comment by danmosemsft** [comment]
@weshaggard the work was done before NS was defined.

**Comment by weshaggard** [comment]
Can they easily be split-up or completely removed then? If so I would suggest we do that. The Sql Xml APIs are very specialized and if the Sql team isn't using them I don't expect anyone will use them.

**Comment by sepidehMS** [comment]
> Can they easily be split-up or completely removed then?

@weshaggard When I tried making these public types in `System.Xml.Xsl.Runtime` internal, many tests started failing as they could not find the types at runtime. I did more investigation and turns out that there is Xslt code that internally loads and makes calls to these types at runtime to generate IL code. Making these internal, caused the IL generation step to fail. 
Looking at the types, I see that they all have the label [EditorBrowsable(EditorBrowsableState.Never)], this convinces me that these types are not supposed to be directly used by developers and are only made public for the internal IL generation. That said, I think we should leave them as is and they shouldn't be exposed in our contracts either.

**Comment by sepidehMS** [comment]
Removed all irrelevant TODOs and SILVERLIGHT #ifdefs. Also ran ILLink on top of my changes and removed whatever was actually dead.

**Comment by danmosemsft** [comment]
I don't think this comment is worth having.
```
                // TODO: this block is never hit by our tests. (Tracked by Issue 16683)

```

**Comment by krwq** [comment]
:shipit:

**Comment by danmosemsft** [comment]
> I did more investigation and turns out that there is Xslt code that internally loads and makes calls to these types at runtime to generate IL code. 

This seems a bit wacky and I think it is worth opening an issue for "future" to figure out why XmlSqlBinaryReader exists. I mean the name itself suggests it shouldn't...

Certainly it's fine to leave right now and move on to other issues

**Comment by weshaggard** [comment]
Thanks for looking into to a simple removal @sepidehMS, if they are needed by xslt then they need to remain.

@danmosemsft I think the name System.Data.SqlXml is a little miss-leading which is why I thought it had to do with SqlXml actually looking at the source the original name was System.Xml.Utils, I'm not sure why it was renamed but it seems to have little to do with SqlXml. 

I'm fine with where we landed given that we can at least explain most of the delta. Thanks for cleaning this up.

**Comment by sepidehMS** [comment]
Can I merge this since OSX builds are busted right now? @danmosemsft

**Comment by danmosemsft** [comment]
I think I'm comfortable merging this witout OSX - there's no OSX specific code as far as I can see.

**Pull request merged**